Support custom params on CR

chmouel opened the pull request:
We are now supporting injecting template variable as params from CR. This let you add some filtering based on payload body condition and event variable.

See documentation in this commit for more description on how it works.

<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] ♽  Run `make test lint` before submitting a PR (ie: with [pre-commit](https://pipelinesascode.com/dev/tools), no need to waste CPU cycle on CI
- [ ] 📖 If you are adding a user facing feature or make a change of the behavior, please verify that you have documented it
- [ ] 🧪 100% coverage is not a target but most of the time we would rather have a unit test if you make a code change.
- [ ] 🎁 If that's something that is possible to do please ensure to check if we can add a e2e test.
- [ ] 🔎 If there is a flakiness in the CI tests then don't *necessary* ignore it, better get the flakyness fixed before merging or if that's not possible there is a good reason to bypass it. (token rate limitation may be a good reason to skip).

chmouel committed:
Support custom params on CR

We are now supporting injecting template variable as params from CR.
This let you add some filtering based on payload body condition and
event variable.

See documentation in this commit for more description on how it works.

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
/retest

chmouel commented:
/test doc-generation

chmouel commented:
/retest

chmouel commented:
/test doc-generation

savitaashture commented on test/pkg/logs/pod.go:
```diff
@@ -11,15 +11,27 @@ import (
-	ns, err := kclient.Namespaces().Get(ctx, "pipelines-as-code", metav1.GetOptions{})
-	if err != nil {
-		ns, err = kclient.Namespaces().Get(ctx, "openshift-pipelines", metav1.GetOptions{})
+func GetControllerLog(ctx context.Context, kclient corev1i.CoreV1Interface, labelselector, containerName string) (string, error) {
```
I am confused with name actually the code is getting ns and passing that ns to GetPodLog 
why we kept name as `GetControllerLog` ?

chmouel replied:
GetControllerLog use GetPodLog, getcontrollerlog tries to figure out if the controller is in openshfit-pipelines or pipelines-as-code namespaces... It used to only be able to see logs form the controller, but i needed that function to read pod logs if the params are properly set and used.... so that's basically a refactor. i think if you read the file directly with no diff it would make things clearer....

savitaashture commented on pkg/pipelineascode/template.go:
```diff
@@ -0,0 +1,181 @@
+			mapFilters[v.Name] = v.Value
+		}
+		// if the name is empty we skip it
+		if v.Name == "" {
```
This should be the first condition before the `v.Filter` check :thinking: because if name is empty we should not do anything right

chmouel replied:
yeah sounds good!

savitaashture commented on docs/content/docs/guide/repositorycrd.md:
```diff
@@ -76,3 +91,115 @@ request with a `concurrency_limit` of 1 in the repository configuration, then al
+
+## Custom Parameter Expansion
+
+Using the `{{ param }}` syntax, Pipelines as Code let yo expand a variable
```
```suggestion
Using the `{{ param }}` syntax, Pipelines as Code let you expand a variable
```

savitaashture commented on docs/content/docs/guide/repositorycrd.md:
```diff
@@ -29,33 +32,45 @@ spec:
-execution to a namespace they don't have access to. To use that feature you
-need to add this annotation to the pipeline annotation :
+This annotation helps prevent bad actors on a cluster from hijacking
+PipelineRun execution to a namespace they don't have access to. It let the user
```
```suggestion
PipelineRun execution to a namespace they don't have access to it. let the user
```

chmouel replied:
heu, i don't think i understand your correction

chmouel committed:
add fix-markdownlint makefile target

will help correct those stupid markdonwlint error we don't really care
about but we care about

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

savitaashture reviewed (approved):
/lgtm

savitaashture commented on pkg/pipelineascode/template.go:
```diff
@@ -0,0 +1,181 @@
+					"ParamsFilterSkipped", fmt.Sprintf("skipping params name %s, filter condition is false", value.Name))
+				continue
+			}
+			mapFilters[value.Name] = value.Value
```
What if value.Value is `" "` ?

I mean its possible that name is given without the value

chmouel replied:
ah yeah, it would be just empty... like the same way if you didn't have anything.. it could be used with filters to set the variable to empty if it match a filter.. maybe i should document that

savitaashture replied:
I would prefer to document it so that there won't be confusions

savitaashture commented on docs/content/docs/guide/repositorycrd.md:
```diff
@@ -78,3 +91,115 @@ request with a `concurrency_limit` of 1 in the repository configuration, then al
+
+{{< hint info >}}
+
+- If you have a `value` and a `secretRef` defined, the `value` will be used.
+- If you don't have a `name` in the `params` the parameter will not parsed.
+- If you have multiple `params` with the same `name` the last one will be used.
```
what happens if name is given but no secret/value/filters

chmouel replied:
same comment as before, it would just not fail and can be used as is.. maybe we should erroring but since it will have the same behaviour as when you don't have params i think it's safe?

savitaashture commented:
/hold
